recirq/qcqmc/for_refactor.py

```diff
@@ -0,0 +1,188 @@
+import abc
+import dataclasses
+import hashlib
+import pathlib
+from dataclasses import dataclass
+from typing import Optional, Sequence, Union
+
+import cirq
+import numpy as np
+
+
+# should go in Dedicated module
+@dataclass(frozen=True, repr=False)
+class Params(abc.ABC):
+    name: str
+
+    def __eq__(self, other):
+        """A helper method to compare two dataclasses which might contain arrays."""
+        if self is other:
+            return True
+
+        if self.__class__ is not other.__class__:
+            return NotImplemented
+
+        return all(
+            array_compatible_eq(thing1, thing2)
+            for thing1, thing2 in zip(
+                dataclasses.astuple(self), dataclasses.astuple(other)
+            )
+        )
+
+    @property
+    def path_string(self) -> str:
+        raise NotImplementedError()
+
+    @property
+    def base_path(self) -> pathlib.Path:
+        return pathlib.Path(self.path_string + "_" + self.hash_key)
+
+    @property
+    def hash_key(self) -> str:
+        """Gets the hash key for a set of params."""
+        return hashlib.md5(str(self).encode("utf-8")).hexdigest()[0:16]
+
+    def __hash__(self):
+        return hash(str(self))
+
+    def __repr__(self):
+        # This custom repr lets us add fields with default values without changing
+        # the repr. This, in turn, lets us use the hash_key reliably even when
+        # we add new fields with a default value.
+        fields = dataclasses.fields(self)
+        # adjusted_fields = [f for f in fields if getattr(self, f.name) != f.default]
+        adjusted_fields = [
+            f
+            for f in fields
+            if not array_compatible_eq(getattr(self, f.name), f.default)
+        ]
+
+        return (
+            self.__class__.__qualname__
+            + "("
+            + ", ".join([f"{f.name}={getattr(self, f.name)}" for f in adjusted_fields])
+            + ")"
+        )
+
+    @classmethod
+    def _json_namespace_(cls):
+        return "recirq.qcqmc"
+
+
+# Should go in dedicated module
+@dataclass(frozen=True, eq=False)
+class Data(abc.ABC):
```

Comment:
would have expected to find this in `data.py` :)

Reply:
Good point.